sql/cagg_api.sql

```diff
@@ -0,0 +1,182 @@
+         ranges AS MATERIALIZED (
+           SELECT width,
+                  tstzrange(@extschema@.time_bucket(width, start),
+                            @extschema@.time_bucket(width, finish - '1 microsecond'::interval) + width) AS bucket
+             FROM timestamps CROSS JOIN UNNEST(bucket_widths) w(width)
+         )
+      SELECT width, pg_current_snapshot()::text, range_agg(bucket) ranges
+        FROM ranges GROUP BY width
+   );
+END
+$body$
+LANGUAGE plpgsql
+SET search_path = pg_catalog, pg_temp;
+
+CREATE OR REPLACE FUNCTION _timescaledb_functions.get_hypertable_invalidations(
+   hypertable REGCLASS,
+   base TIMESTAMP,
+   bucket_widths INTERVAL[]
+) RETURNS TABLE (bucket_width INTERVAL, token TEXT, invalidations TSMULTIRANGE) AS
+$body$
+DECLARE
+   l_hypertable_id INTEGER := _timescaledb_functions.get_hypertable_id(hypertable, 'timestamp'::regtype);
+BEGIN
+   RETURN QUERY (
+      WITH
+         -- Collect ranges from the invalidation log and convert them
+         -- to correct type.
+         timestamps AS MATERIALIZED (
+            SELECT _timescaledb_functions.to_timestamp_without_timezone(lowest_modified_value) AS start,
+                   _timescaledb_functions.to_timestamp_without_timezone(greatest_modified_value) AS finish
+              FROM _timescaledb_catalog.continuous_aggs_hypertable_invalidation_log
+             WHERE hypertable_id = l_hypertable_id
+         ),
+         -- Compute bucket-aligned ranges from the ranges above. Since
+         -- a range can end at the end of a bucket and this range
+         -- should not include the next bucket, we need to subtract
+         -- one microsecond before computing the bucket and then add
+         -- the width again.
+         ranges AS MATERIALIZED (
+           SELECT width,
+                  tsrange(@extschema@.time_bucket(width, start),
+                          @extschema@.time_bucket(width, finish - '1 microsecond'::interval) + width) AS bucket
+             FROM timestamps CROSS JOIN UNNEST(bucket_widths) w(width)
+         )
+      SELECT width, pg_current_snapshot()::text, range_agg(bucket)
+        FROM ranges GROUP BY width
+   );
+END
+$body$
+LANGUAGE plpgsql
+SET search_path TO pg_catalog, pg_temp;
+
+-- Accept a set of hypertable invalidations for a hypertable.
+--
+-- This procedure is used to accept all invalidations for a hypertable
+-- using the token returned by a previous call of
+-- get_hypertable_invalidations().
+CREATE OR REPLACE PROCEDURE _timescaledb_functions.accept_hypertable_invalidations(
+   hypertable REGCLASS,
+   token TEXT
+) AS
+$body$
+DECLARE
+   info RECORD;
+   errmsg TEXT;
+BEGIN
+   SELECT ht.id AS hypertable_id,
+          EXISTS(
+	     SELECT
+	       FROM _timescaledb_catalog.continuous_agg
+	      WHERE raw_hypertable_id = ht.id
+	  ) AS has_cagg
+     INTO info
+     FROM _timescaledb_catalog.hypertable ht
+    WHERE format('%I.%I', schema_name, table_name)::regclass = hypertable;
```

Comment:
```suggestion
   SELECT ht.id AS hypertable_id,
          (cagg.raw_hypertable_id IS NOT NULL) AS has_cagg
     INTO info
     FROM _timescaledb_catalog.hypertable ht
     LEFT JOIN _timescaledb_catalog.continuous_agg cagg ON cagg.raw_hypertable_id = ht.id
    WHERE format('%I.%I', schema_name, table_name)::regclass = hypertable;
```
Nit. This is more a matter of taste!

Reply:
Hmmm not sure if in this case that you're executing the EXISTS for each row returned by the query is faster than the left join

Reply:
As I was suspecting:
```sql
fabrizio=# CREATE TABLE foo AS SELECT i FROM generate_series(1,1000000) AS i;
SELECT 1000000
fabrizio=# CREATE INDEX foo_idx ON foo(i);
CREATE INDEX
fabrizio=# CREATE TABLE bar AS SELECT i FROM generate_series(1,1000000) AS i WHERE (i%2)=0;
SELECT 500000
fabrizio=# CREATE INDEX bar_idx ON bar(i);
CREATE INDEX
fabrizio=# VACUUM (ANALYZE);
VACUUM
fabrizio=# EXPLAIN (ANALYZE, BUFFERS) SELECT foo.i, bar.i IS NOT NULL AS has_bar FROM foo LEFT JOIN bar ON bar.i = foo.i;
                                                                QUERY PLAN                                                                
------------------------------------------------------------------------------------------------------------------------------------------
 Merge Left Join  (cost=1.25..35408.79 rows=1000000 width=5) (actual time=0.105..489.541 rows=1000000 loops=1)
   Merge Cond: (foo.i = bar.i)
   Buffers: shared hit=10 read=4096
   ->  Index Only Scan using foo_idx on foo  (cost=0.42..17750.42 rows=1000000 width=4) (actual time=0.053..175.225 rows=1000000 loops=1)
         Heap Fetches: 0
         Buffers: shared hit=5 read=2731
   ->  Index Only Scan using bar_idx on bar  (cost=0.42..8909.42 rows=500000 width=4) (actual time=0.044..89.034 rows=500000 loops=1)
         Heap Fetches: 0
         Buffers: shared hit=5 read=1365
 Planning:
   Buffers: shared hit=126 read=10
 Planning Time: 2.489 ms
 Execution Time: 527.760 ms
(13 rows)

fabrizio=# EXPLAIN (ANALYZE, BUFFERS) SELECT foo.i, EXISTS (SELECT FROM bar WHERE bar.i = foo.i) AS has_bar FROM foo;
                                                           QUERY PLAN                                                           
--------------------------------------------------------------------------------------------------------------------------------
 Seq Scan on foo  (cost=0.00..2454430.00 rows=1000000 width=5) (actual time=49.469..4259.501 rows=1000000 loops=1)
   Buffers: shared hit=3002366 read=2065
   SubPlan 1
     ->  Index Only Scan using bar_idx on bar  (cost=0.42..2.44 rows=1 width=0) (actual time=0.004..0.004 rows=0 loops=1000000)
           Index Cond: (i = foo.i)
           Heap Fetches: 0
           Buffers: shared hit=2999998 read=3
 Planning:
   Buffers: shared hit=8
 Planning Time: 0.620 ms
 JIT:
   Functions: 7
   Options: Inlining true, Optimization true, Expressions true, Deforming true
   Timing: Generation 0.144 ms (Deform 0.037 ms), Inlining 34.323 ms, Optimization 8.801 ms, Emission 6.274 ms, Total 49.542 ms
 Execution Time: 4344.775 ms
(15 rows)
```

Ok then JIT can be the problem so let's turn it off:
```sql
fabrizio=# SET jit=off;
SET
fabrizio=# EXPLAIN (ANALYZE, BUFFERS) SELECT foo.i, bar.i IS NOT NULL AS has_bar FROM foo LEFT JOIN bar ON bar.i = foo.i;
                                                                QUERY PLAN                                                                
------------------------------------------------------------------------------------------------------------------------------------------
 Merge Left Join  (cost=1.25..35408.79 rows=1000000 width=5) (actual time=0.100..459.251 rows=1000000 loops=1)
   Merge Cond: (foo.i = bar.i)
   Buffers: shared hit=4106
   ->  Index Only Scan using foo_idx on foo  (cost=0.42..17750.42 rows=1000000 width=4) (actual time=0.048..155.367 rows=1000000 loops=1)
         Heap Fetches: 0
         Buffers: shared hit=2736
   ->  Index Only Scan using bar_idx on bar  (cost=0.42..8909.42 rows=500000 width=4) (actual time=0.043..80.370 rows=500000 loops=1)
         Heap Fetches: 0
         Buffers: shared hit=1370
 Planning:
   Buffers: shared hit=24
 Planning Time: 0.894 ms
 Execution Time: 498.252 ms
(13 rows)

fabrizio=# EXPLAIN (ANALYZE, BUFFERS) SELECT foo.i, EXISTS (SELECT FROM bar WHERE bar.i = foo.i) AS has_bar FROM foo;
                                                           QUERY PLAN                                                           
--------------------------------------------------------------------------------------------------------------------------------
 Seq Scan on foo  (cost=0.00..2454430.00 rows=1000000 width=5) (actual time=0.150..4212.894 rows=1000000 loops=1)
   Buffers: shared hit=3002401 read=2030
   SubPlan 1
     ->  Index Only Scan using bar_idx on bar  (cost=0.42..2.44 rows=1 width=0) (actual time=0.004..0.004 rows=0 loops=1000000)
           Index Cond: (i = foo.i)
           Heap Fetches: 0
           Buffers: shared hit=3000001
 Planning:
   Buffers: shared hit=8
 Planning Time: 0.522 ms
 Execution Time: 4291.370 ms
(11 rows)
```

Reply:
Oh, big difference. Thanks for making the tests.